Add option to return matched/unmatched/unexpected in 'load_state_dict' fixed #9530

bhushan23 opened the pull request:
Add option to return matched/unmatched/unexpected in 'load_state_dict'

Adding test case in TestNN.test_load_state_dict 
Implements #9530

bhushan23 commented:
builds are failing with following error
In file included from /var/lib/jenkins/workspace/torch/../aten/src/ATen/cuda/CUDAContext.h:4:0,
02:05:03                  from /var/lib/jenkins/workspace/torch/csrc/cuda/comm.h:5,
02:05:03                  from /var/lib/jenkins/workspace/torch/csrc/autograd/functions/comm.cpp:6:
02:05:03 /var/lib/jenkins/workspace/torch/../aten/src/ATen/cuda/CUDAStream.h:6:30: fatal error: cuda_runtime_api.h: No such file or directory
02:05:03  #include "cuda_runtime_api.h"

ezyang commented:
@pytorchbot retest this please

bhushan23 commented:
Added test is asserting in one of the build. Need to update the tests, will do that today but this does not blocks code review of load_state

apaszke commented:
Uh I missed the comments in the original issue. Let's finalize the discussion on the interface before we merge this.

apaszke reviewed (changes requested):
Didn't take a look at the code yet. We should decide on the API first.

bhushan23 committed:
Introducing diff_state_dict -
Returns matched, unmatched and unexpected keys from given
state dictionary

test test_diff_state_dict added

reviewed by @fmassa

returning keys as set instead of list

Using Dictionary for returning keys

bhushan23 commented:
@fmassa please review the prototype 
Please don't do in depth code review. This is just to verify and confirm approach

bhushan23 commented:
@fmassa @soumith can you review the prototype?

yf225 commented:
@fmassa @soumith Pinging for prototype review

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -700,28 +710,19 @@ def load_state_dict(self, state_dict, strict=True):
         if len(error_msgs) > 0:
             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
                                self.__class__.__name__, "\n\t".join(error_msgs)))
+        if return_keys:
```
I think we have decided that it's better not to return the keys conditionally on an argument?

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -687,9 +681,25 @@ def load_state_dict(self, state_dict, strict=True):
         error_msgs = []
+        
+        if strict or return_keys:
+            matched_keys, missing_keys, unexpected_keys = self.diff_state_dict(state_dict)
```
Can't we always compute this?

fmassa reviewed (changes requested):
I think this looks ok, but from what I understood we would be avoiding returning the `matched` / `unmatched` / `unexpected` ?

bhushan23 commented:
@fmassa you are right, removing returning keys from load_state_dict

Regarding, computing keys always, we don't need to.
We will compute keys only if strict is passed

bhushan23 commented:
this is weird, checking for difference before loading leads to some unmatched keys but doing within load and after does not. @fmassa do you know why? 
Have updated change with this.

bhushan23 commented:
@apaszke are you also reviewing?

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -722,6 +724,10 @@ def load(module, prefix=''):
             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
                                self.__class__.__name__, "\n\t".join(error_msgs)))
 
+        if len(error_msgs) > 0:
```
looks like this block is duplicated?

fmassa commented on torch/jit/__init__.py:
```diff
@@ -868,7 +868,7 @@ def _get_methods(cls):
     'forward', 'register_buffer', 'register_parameter', 'add_module',
     '_apply', 'apply', 'cuda', 'cpu', 'type', 'float', 'double', 'half',
-    'state_dict', 'load_state_dict', '_load_from_state_dict',
+    'state_dict', 'load_state_dict', 'diff_state_dict', '_load_from_state_dict',
```
@zou3519 does the JIT know how to compile functions using python `set`?

zou3519 replied:
What do you mean by "compile functions using python set"?

fmassa replied:
I probably don't know what I'm saying, but what's the purpose of `_compiled_methods_whitelist` in the JIT?

fmassa reviewed (changes requested):
There is a duplicated error message that should be fixed.
Also, I'm not sure that the JIT supports that function, but I'll let the others comment on that.

Apart from that, it generally looks good to me

bhushan23 commented:
@pytorchbot retest

bhushan23 commented:
@pytorchbot retest

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -663,16 +656,27 @@ def _load_from_state_dict(self, state_dict, prefix, metadata, strict, missing_ke
+        # unexpected keys are present in state but not in current
+        unexpected_keys = state_keys - local_keys
+
+        return list(matched_keys), list(unmatched_keys), list(unexpected_keys)
```
I'm not 100% sure if we really want to have this function. The beginning could be simplified to
```py
local_keys = set(self.state_dict())
state_keys = set(state_dict)
```
and then it becomes a few lines of very straightforward Python code. To be honest it would be harder for me to find this function to the docs, and make sure that I assign the three lists we return in the correct order, than to write them out myself...

apaszke reviewed (changes requested):
I have serious doubts about this functionality. It doesn't seem worth having in the core for me. Everyone can implement them in their projects in case they need it, and it might even end up being faster in case you don't need one of the three sets we compute.

bhushan23 commented:
@apaszke 
>> Everyone can implement them in their projects in case they need it, 
This this API, one can quickly query state of the state_dict (and one get full picture)
And doing this in code is a bit messy, why should we force everyone to find the difference keys manually every time they need it?
>> and it might even end up being faster in case you don't need one of the three sets we compute.
I agree here. 
How about passing keys of interest and we only compute those?
But, I guess, if someone is using diff_state_dict, they are okay with extra time.

Also, now returning keys as set instead of converting into list. 
Keys are supposed to be set, then why waste time in un-necessary conversion into list.

apaszke commented:
I'm just not convinced that it's a good API. It's a very simple thing, and we're making the interface of a module unnecessarily complex. Also, as I mentioned, returning 3 values of the same type makes it extremely easy to mix them up, and there's really no way to remember what this function does without looking it up in the docs every time. On the other hand, if you had this code in your project, it would be much harder to make a mistake.

bhushan23 commented:
@apaszke I get your point. Returning 3 different keys can lead to confusion. 
How about returning dictionary? Key will be string (matched / unmatched / unexpected) and value will be set of those keys.

fmassa commented:
After a lot of discussion, we think that it's better to let the user handle this manually, and use `strict=True` mode when loading.

Thanks a lot for this PR!
